src/tribol/mesh/CouplingScheme.cpp

```diff
@@ -1272,98 +1323,95 @@ void CouplingScheme::computeTimeStep(real &dt)
-      // the pair is not a contact candidate
-      if (!pair.isContactCandidate)
-      {
-         continue;
-      }
-
-      real x1[dim * numNodesPerCell1];
-      real v1[dim * numNodesPerCell1];
-      mesh1.getFaceCoords( pair.pairIndex1, &x1[0] );
-      mesh1.getFaceNodalVelocities( pair.pairIndex1, &v1[0] );
-
-      real x2[dim * numNodesPerCell2];
-      real v2[dim * numNodesPerCell2];
-      mesh2.getFaceCoords( pair.pairIndex2, &x2[0] );
-      mesh2.getFaceNodalVelocities( pair.pairIndex2, &v2[0] );
+  // note: the timestep vote is based on a maximum allowable interpenetration
+  // approach checking current gaps and then performing a velocity projection.
+  // This timestep vote is not derived from a stability analysis and does not 
+  // account for the spring stiffness in a CFL-like timestep constraint. 
+  // The timestep vote in this routine is intended to avoid missing contact or 
+  // inadequately resolving contact by detecting 'too much' interpenetration. 
+  // To do this, the first check is a 'gap check' where the current gap is 
+  // checked against a fraction of the element-thicknesses. The second check 
+  // assumes zero gap, and then does a velocity projection followed by the 
+  // 'gap check' using this new, projected gap.
+
+  auto& mesh1 = getMesh1();
+  auto& mesh2 = getMesh2();
+ 
+  // check if the element thicknesses have been set. This is now
+  // regardless of exact penalty calculation type, since element
+  // thicknesses are required for auto contact even if a constant
+  // penalty is used
+  if ( !mesh1.getElementData().m_is_element_thickness_set ||
+       !mesh2.getElementData().m_is_element_thickness_set )
+  {
+    return; 
+  }
+
+  RealT proj_ratio = m_parameters.timestep_pen_frac;
+  //int num_sides = 2; // always 2 sides in a single coupling scheme
+  int dim = spatialDimension();
+
+  // loop over each interface pair. Even if pair is not in contact, 
+  // we still do a velocity projection for that proximate face-pair 
+  // to see if interpenetration next cycle 'may' be too much
+  auto cs_view = getView();
+  ArrayT<RealT> dt_temp_data({dt, dt}, getAllocatorId());
+  ArrayViewT<RealT> dt_temp = dt_temp_data;
+  // [0]: exceed_max_gap1, [1]: exceed_max_gap2, [2]: neg_dt_gap_msg, [3]: neg_dt_vel_proj_msg
+  ArrayT<bool> msg_data({false, false, false, false}, getAllocatorId());
+  ArrayViewT<bool> msg = msg_data;
+  forAllExec(getExecutionMode(), getNumActivePairs(),
+    [cs_view, dim, proj_ratio, msg, dt_temp, dt] TRIBOL_HOST_DEVICE (IndexT i)
+    {
+      auto& plane = cs_view.getContactPlane(i);
+
+      auto& mesh1 = cs_view.getMesh1();
+      auto& mesh2 = cs_view.getMesh2();
+
+      // get pair indices
+      IndexT index1 = plane.getCpElementId1();
+      IndexT index2 = plane.getCpElementId2();
+
+      constexpr int max_dim = 3;
+      constexpr int max_nodes_per_elem = 4;
+      StackArrayT<RealT, max_dim * max_nodes_per_elem> x1;
```

Comment:
Here you use a StackArrayT. Make sure to comment in the array types when/why you'd use this over one of the other arrays.